src/pal/src/numa/numa.cpp

```diff
@@ -0,0 +1,683 @@
+    memset(g_groupAndIndexToCpu, 0xff, g_groupCount * MaxCpusPerGroup * sizeof(short));
+    memset(g_cpuToAffinity, 0xff, g_cpuCount * sizeof(CpuAffinity));
+    memset(g_groupToCpuMask, 0, g_groupCount * sizeof(KAFFINITY));
+    memset(g_groupToCpuCount, 0, g_groupCount * sizeof(BYTE));
+}
+
+/*++
+Function:
+  FreeLookupArrays
+
+Free CPU and group lookup arrays
+--*/
+VOID
+FreeLookupArrays()
+{
+    free(g_groupAndIndexToCpu);
+    free(g_cpuToAffinity);
+    free(g_groupToCpuMask);
+    free(g_groupToCpuCount);
+
+    g_groupAndIndexToCpu = NULL;
+    g_cpuToAffinity = NULL;
+    g_groupToCpuMask = NULL;
+    g_groupToCpuCount = NULL;
+}
+
+/*++
+Function:
+  GetFullAffinityMask
+
+Get affinity mask for the specified number of processors with all
+the processors enabled.
+--*/
+KAFFINITY GetFullAffinityMask(int cpuCount)
+{
+    return ((KAFFINITY)1 << (cpuCount)) - 1;
+}
+
+/*++
+Function:
+  NUMASupportInitialize
+
+Initialize data structures for getting and setting thread affinities to processors and
+querying NUMA related processor information.
+On systems with no NUMA support, it behaves as if there was a single NUMA node with
+a single group of processors.
+--*/
+BOOL
+NUMASupportInitialize()
+{
+#if HAVE_NUMA_H
+    if (numa_available() != -1)
+    {
+        struct bitmask *mask = numa_allocate_cpumask();
+        int numaNodesCount = numa_max_node() + 1;
+
+        g_cpuCount = numa_num_possible_cpus();
+        g_groupCount = 0;
+
+        for (int i = 0; i < numaNodesCount; i++)
+        {
+            int st = numa_node_to_cpus(i, mask);
+            unsigned int nodeCpuCount = numa_bitmask_weight(mask);
+            unsigned int nodeGroupCount = (nodeCpuCount + MaxCpusPerGroup - 1) / MaxCpusPerGroup;
+            g_groupCount += nodeGroupCount;
+        }
+
+        AllocateLookupArrays();
+
+        WORD currentGroup = 0;
+        int currentGroupCpus = 0;
+
+        for (int i = 0; i < numaNodesCount; i++)
+        {
+            int st = numa_node_to_cpus(i, mask);
+            unsigned int nodeCpuCount = numa_bitmask_weight(mask);
+            unsigned int nodeGroupCount = (nodeCpuCount + MaxCpusPerGroup - 1) / MaxCpusPerGroup;
+            for (int j = 0; j < g_cpuCount; j++)
+            {
+                if (numa_bitmask_isbitset(mask, j))
+                {
+                    if (currentGroupCpus == MaxCpusPerGroup)
+                    {
+                        currentGroupCpus = 0;
+                        currentGroup++;
+                    }
+                    g_cpuToAffinity[j].Node = i;
+                    g_cpuToAffinity[j].Group = currentGroup;
+                    g_cpuToAffinity[j].Number = currentGroupCpus;
+                    g_groupAndIndexToCpu[currentGroup * MaxCpusPerGroup + currentGroupCpus] = j;
+                    currentGroupCpus++;
+                }
+            }
+
+            if (currentGroupCpus != 0)
+            {
+                g_groupToCpuCount[currentGroup] = currentGroupCpus;
```

Comment:
`g_groupToCpuCount[currentGroup]` is set here for whatever `currentGroup` is when we get here. What if we filled a group and moved on to the next one in the for loop above (near line 175)?